crates/air-lib/air-parser/src/parser/lexer/call_variable_parser.rs

```diff
@@ -273,59 +272,57 @@ impl<'input> CallVariableParser<'input> {
         self.current_pos() == self.string_to_parse.len() - 1
     }
 
-    fn to_variable<'v>(&self, variable_name: &'v str) -> AstVariable<'v> {
+    fn to_variable_token<'v>(&self, name: &'v str) -> Token<'v> {
         if self.state.is_first_stream_tag {
-            // TODO: cut the stream tag after the refactoring.
-            AstVariable::Stream(variable_name)
+            Token::Stream { name }
         } else {
-            AstVariable::Scalar(variable_name)
+            Token::Scalar { name }
         }
     }
 
-    fn try_to_variable_and_lambda(
-        &self,
-        pos: usize,
-    ) -> LexerResult<(&'input str, LambdaAST<'input>)> {
+    fn to_variable_token_with_lambda<'v>(&self, name: &'v str, lambda: LambdaAST<'v>) -> Token<'v> {
+        if self.state.is_first_stream_tag {
+            Token::StreamWithLambda { name, lambda }
+        } else {
+            Token::ScalarWithLambda { name, lambda }
+        }
+    }
+
+    fn try_to_variable_and_lambda(&self, lambda_start_pos: usize) -> LexerResult<Token<'input>> {
         // +2 to ignore ".$" prefix
-        let lambda = crate::parse_lambda(&self.string_to_parse[pos + 2..]).map_err(|e| {
-            LexerError::LambdaParserError(
-                self.start_pos + pos,
-                self.start_pos + self.string_to_parse.len(),
-                e.to_string(),
-            )
-        })?;
-
-        Ok((&self.string_to_parse[0..pos], lambda))
+        let lambda =
+            crate::parse_lambda(&self.string_to_parse[lambda_start_pos + 2..]).map_err(|e| {
+                LexerError::LambdaParserError(
+                    self.start_pos + lambda_start_pos,
+                    self.start_pos + self.string_to_parse.len(),
+                    e.to_string(),
```

Comment:
It seems that's the only place where you instantiate `LexerError::LambdaParserError`, maybe no need for `to_string` and you can move the whole `air_lambda_parser::LambdaParserError` to the enum variant?

Reply:
It's under [TODO](https://github.com/fluencelabs/aquavm/blob/1c55d34981d86132bb10c205bf14268320bc1a75/crates/air-lib/air-parser/src/parser/lexer/errors.rs#L51) already :). It'll require introducing a lifetime and a some work to propagete it...